File: Sources/Domain/Models/UsageQuota.swift
import Foundation

/// Represents a single usage quota measurement for an AI provider.
/// This is a rich domain model that encapsulates quota-related behavior.
public struct UsageQuota: Sendable, Equatable, Hashable, Comparable {
    /// The percentage of quota remaining (0-100)
    public let percentRemaining: Double

    /// The type of quota (session, weekly, model-specific)
    public let quotaType: QuotaType

    /// The AI provider this quota belongs to
    public let provider: AIProvider

    /// When this quota will reset (if known)
    public let resetsAt: Date?

    // MARK: - Initialization

    public init(
        percentRemaining: Double,
        quotaType: QuotaType,
        provider: AIProvider,
        resetsAt: Date? = nil
    ) {
        self.percentRemaining = max(0, min(100, percentRemaining))
        self.quotaType = quotaType
        self.provider = provider
        self.resetsAt = resetsAt
    }

    // MARK: - Domain Behavior

    /// The current health status based on percentage remaining.
    /// This is a domain rule: status is determined by business thresholds.
    public var status: QuotaStatus {
        QuotaStatus.from(percentRemaining: percentRemaining)
    }

    /// The percentage that has been used (0-100)
    public var percentUsed: Double {
        100 - percentRemaining
    }

    /// Whether this quota is completely exhausted
    public var isDepleted: Bool {
        percentRemaining <= 0
    }

    /// Whether this quota needs attention (warning, critical, or depleted)
    public var needsAttention: Bool {
        status.needsAttention
    }

    /// Time until this quota resets (if known)
    public var timeUntilReset: TimeInterval? {
        guard let resetsAt else { return nil }
        return max(0, resetsAt.timeIntervalSinceNow)
    }

    /// Human-readable description of time until reset
    public var resetDescription: String? {
        guard let timeUntilReset else { return nil }

        let hours = Int(timeUntilReset / 3600)
        let minutes = Int((timeUntilReset.truncatingRemainder(dividingBy: 3600)) / 60)

        if hours > 24 {
            let days = hours / 24
            return "Resets in \(days)d \(hours % 24)h"
        } else if hours > 0 {
            return "Resets in \(hours)h \(minutes)m"
        } else if minutes > 0 {
            return "Resets in \(minutes)m"
        } else {
            return "Resets soon"
        }
    }

    // MARK: - Comparable

    public static func < (lhs: UsageQuota, rhs: UsageQuota) -> Bool {
        lhs.percentRemaining < rhs.percentRemaining
    }
}

File: shared/domain/src/commonMain/kotlin/com/tddworks/claudebar/domain/model/UsageQuota.kt
package com.tddworks.claudebar.domain.model

import kotlinx.datetime.Clock
import kotlinx.datetime.Instant
import kotlin.time.Duration
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.minutes

/**
 * Represents a single usage quota measurement for an AI provider.
 * This is a rich domain model that encapsulates quota-related behavior.
 */
data class UsageQuota(
    /** The percentage of quota remaining (0-100) */
    val percentRemaining: Double,

    /** The type of quota (session, weekly, model-specific) */
    val quotaType: QuotaType,

    /** The provider ID this quota belongs to (e.g., "claude", "codex", "gemini") */
    val providerId: String,

    /** When this quota will reset (if known) */
    val resetsAt: Instant? = null,

    /** Raw reset text from CLI (e.g., "Resets 11am", "Resets Jan 15") */
    val resetText: String? = null
) : Comparable<UsageQuota> {

    init {
        require(percentRemaining in 0.0..100.0) {
            "percentRemaining must be between 0 and 100"
        }
    }

    /**
     * The current health status based on percentage remaining.
     * This is a domain rule: status is determined by business thresholds.
     */
    val status: QuotaStatus
        get() = QuotaStatus.from(percentRemaining)

    /** The percentage that has been used (0-100) */
    val percentUsed: Double
        get() = 100 - percentRemaining

    /** Whether this quota is completely exhausted */
    val isDepleted: Boolean
        get() = percentRemaining <= 0

    /** Whether this quota needs attention (warning, critical, or depleted) */
    val needsAttention: Boolean
        get() = status.needsAttention

    /** Time until this quota resets (if known) */
    val timeUntilReset: Duration?
        get() = resetsAt?.let { resetTime ->
            val now = Clock.System.now()
            val diff = resetTime - now
            if (diff.isNegative()) Duration.ZERO else diff
        }

    /** Human-readable description of time until reset */
    val resetDescription: String?
        get() = timeUntilReset?.let { duration ->
            val hours = duration.inWholeHours
            val minutes = (duration.inWholeMinutes % 60).toInt()

            when {
                hours > 24 -> {
                    val days = hours / 24
                    "Resets in ${days}d ${hours % 24}h"
                }
                hours > 0 -> "Resets in ${hours}h ${minutes}m"
                minutes > 0 -> "Resets in ${minutes}m"
                else -> "Resets soon"
            }
        }

    override fun compareTo(other: UsageQuota): Int =
        percentRemaining.compareTo(other.percentRemaining)

    companion object {
        /** Creates a UsageQuota with clamped percentage value */
        fun create(
            percentRemaining: Double,
            quotaType: QuotaType,
            providerId: String,
            resetsAt: Instant? = null,
            resetText: String? = null
        ): UsageQuota = UsageQuota(
            percentRemaining = percentRemaining.coerceIn(0.0, 100.0),
            quotaType = quotaType,
            providerId = providerId,
            resetsAt = resetsAt,
            resetText = resetText
        )
    }
}
